# Plans for 2020

This is a community space for brainstorming ideas for the next BDS Hackathon.

## Recap of 2019

We set out to develop a Python package that provides new analytical functionality for single-cell multi-omics datasets. We created a new object class called MultiAnnData for storing and manipulating AnnData objects of multiple omics modalities. Once the functionality of our module seems sufficient, we will release a public API on PyPI under package name MoPy (pronounced mow-pie) or MadPy (if the latter becomes available).

Thank you, Nathan Sheffield and Jason Papin, for helping make the Hackathon a success !!

## Things to keep

* Use of polling to determine theme/dataset

* Recruiting past BDS trainees to be involved in planning and executing the event (Thanks Greg and Jeff!)

* Use of a single language (i.e. Python)

* Physical proximity in reserved workroom

* Meals (Thanks Kim!)

* Use of GitHub for versioning and sharing of code and ideas

* Focus on emerging genomics technologies

* BDSHackathon computing group on Rivanna

* discussion of pickling

## Things to optimize

* Advanced determination and distribution of a single technology, research paper, and associated dataset (e.g. dataset in shared directory with read access for group)

* Provision or verification of a reliable computing environment in which to immediately (or at least quickly) get started; two parts:

  * Part I: Creation or identification of comprehensive set of modules or environments on Rivanna (or other HPC such as AWS) needed for all phases of project development; may require something as complicated as building a container or as simple as writing a `.bashrc` file.

  * Part II: Dedication of first hour of hacking to verification of successful implementation of programming environment prepared in Part I on each individual participant's system. Might be useful to have a live, step-by-step tutorial for connection to a remote server; confirmation of group membership, permissions, and data access; cloning of BDSHackathon GitHub repository; pushing an initial commit (e.g. README.md per individual); launch of Python (or default IDE); and generation of first plot.

* White board, dry-erase markers, and eraser for recording ideas and decisions

## Dataset proposals 2020

* Long-read sequencing
* Spatial expression
  * mass cytometry based https://www.ionpath.com/publications/
  * sequencing based https://spatialtranscriptomics.com/publications/
